core/trino-main/src/main/java/io/trino/operator/output/PositionsAppender.java

```diff
@@ -11,17 +11,27 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
 package io.trino.operator.output;
 
 import io.trino.spi.block.Block;
-import io.trino.spi.block.BlockBuilder;
-import io.trino.spi.type.Type;
+import io.trino.spi.block.BlockBuilderStatus;
 import it.unimi.dsi.fastutil.ints.IntArrayList;
 
-import static java.util.Objects.requireNonNull;
+import javax.annotation.Nullable;
 
 public interface PositionsAppender
 {
-    void appendTo(IntArrayList positions, Block source, BlockBuilder target);
+    void append(IntArrayList positions, Block source);
+
+    void appendRow(Block source, int position);
+
+    Block build();
+
+    /**
+     * Creates new empty {@link PositionsAppender} of the same type as {@code this}
+     * pre-sized to the current position count and size in bytes.
+     */
+    PositionsAppender newStateLike(@Nullable BlockBuilderStatus blockBuilderStatus);
```

Comment:
After some thought I came to the conclusion that this name is bad. We are not creating any "state" here but a new instance. `newApenderLike` is far from perfect but probably a better one.

Reply:
I don't think we need to create a new instance at all. It's overly complicated (especially if appenders form a hierarchy). You can just reset internal appender state after during `build()` method call.

Reply:
One issue with `reset` is that then we have to worry about the lifecycle of the created Block because it will reuse the underlying primitive arrays from the appenders. In this PR, `Blocks` are immediately serialized so it's ok but it can create subtle bugs if we use this in other places.
A second issue is that `reset` means the underlying arrays never `shrinks` so there are possible pathological cases with skewed hash where we use 2X more memory.